[wasm] Remove System.Reflection.MetadataLoadContext from WasmAppBuilder

mdh1418 opened the pull request:
In the effort to port over samples that demonstrate building on top of the mono runtime to `dotnet/samples`, it is desirable to reduce the amount of dependencies in the sample. 

`System.Reflection.MetadataLoadContext` in `WasmAppBuilder` is utilized to load assemblies (with the intention to not run them) and also load their referenced assemblies. By doing so, a list of assemblies the wasm app depends on is generated and incorporated into `mono-config.js`. 

Using `dotnet publish` in `dotnet/samples` will utilize a linker, which already determines the necessary assemblies required to run the application, removing the reliance on System.Reflection.MetadataLoadContext.

This PR achieves the following:
1. Separates the assembly and reference assembly loading process from the `WasmAppBuilder` into a separate task in `WasmLoadAssembliesAndReferences.cs`
2. Refactors that loading process to take in assemblies, search paths, and the option to skip missing referenced assemblies, thereby reducing the original complexity with respect to `MainAssembly`, `ExtraAssemblies` and having at least one of `Assemblies` or `AssemblySearchPaths`
3. Applies the updated `WasmAppBuilder` and `WasmLoadAssembliesAndReferences` tasks to the library tests (`tests.mobile.targets`), browser sample (`WasmSample.csproj`), debugger-tests (`debugger-tests.csproj`), and the runtime tests (`WasmTestRunner.proj`)

Note: In the libraries test, `tests.mobile.targets`, `WasmAppBuilder` originally resolved and loaded the `MainAssembly` using paths provided in `AssemblySearchPaths` to build the WasmApp. However, the libraries test now uses publish, which does that already. So we now include the assemblies resulting from publish directly into `WasmAppBuilder` in order to run the libraries test. The debugger-tests and runtime tests do not use publish yet, so they will use the new task `WasmLoadAssembliesAndReferences` to resolve necessary assemblies.

Commit:
Remove MetadataLoadContext from WasmAppBuilder

Commit:
Remove assembly loading from WasmAppBuilder

naricc commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAppBuilder.cs:
```diff
@@ -25,10 +25,6 @@ public class WasmAppBuilder : Task
 
-    // If true, continue when a referenced assembly cannot be found.
-    // If false, throw an exception.
-    public bool SkipMissingAssemblies { get; set; }
```
So the thing that is going to have to change in the runtime test is where it uses this parameter: https://github.com/dotnet/runtime/blob/4730697ad2c218f68dad1f5e788a897ac4fb01ea/src/tests/Common/wasm-test-runner/WasmTestRunner.proj#L35

Assuming this still gets all the assemblies and puts them in the right place, I don't think other changes are required for the wasm tests.

steveisok replied:
I added a task that preserves what you need for the runtime tests.  That way we can remove using MDLC from WasmAppBuilder.

naricc replied:
Ok, looks good to me as far as the runtime tests go.

vargaz commented:
Note that in the future the app builder or the build process will also include the PInvokeTableGenerator etc. class which also depends on this assembly, and that dependency is much harder to remove.

Commit:
Add a separate task to determine the referenced assemblies for wasm.  This is so we don't break the runtime tests / require significant changes

akoeplinger commented:
@vargaz that's fine, it's not really the reference to System.Reflection.MetadataLoadContext that's problematic but rather that it was used to duplicate the unused assembly trimming that the linker already does now.

Commit:
Fix wasm libraries test and remove AssemblySearchPaths

steveisok commented on eng/testing/tests.mobile.targets:
```diff
@@ -143,23 +143,27 @@
              AssemblyFile="$(WasmAppBuilderTasksAssemblyPath)" />
 
   <Target Condition="'$(TargetOS)' == 'Browser'" Name="BundleTestWasmApp">
+    <PropertyGroup>
```
This is already done at https://github.com/dotnet/runtime/blob/9765835057496db6ec2c1e5fb7a7610f628915ef/eng/testing/tests.props#L31

steveisok commented:
@mdh1418 The satellite assembly failures are probably related to your changes.  You can reproduce locally.

mdh1418 commented:
With these changes, there is now another assembly added to `artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/AppBundle/managed`
After this PR
```
mdhwang:~/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/AppBundle/managed$ find . -name 'System.Runtime.Loader.Tests.resources.dll'
./es-MX/System.Runtime.Loader.Tests.resources.dll
./en/System.Runtime.Loader.Tests.resources.dll
./System.Runtime.Loader.Tests.resources.dll
```
Before this PR
```
mdhwang:~/runtime/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/AppBundle/managed$ find . -name 'System.Runtime.Loader.Tests.resources.dll'
./es-MX/System.Runtime.Loader.Tests.resources.dll
./en/System.Runtime.Loader.Tests.resources.dll
```

In both test runs, before and after this PR, https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/mono/mono/metadata/assembly.c#L2789 is being hit, which results in `is_satellite` being evaluated as `False` https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/mono/mono/metadata/assembly.c#L2623. 
The condition https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/mono/mono/metadata/assembly.c#L2581 is never met in the passing case, and is reached in the failing case. (138 assemblies checked in the passing case, 224+ assemblies checked in the failing case)

@CoffeeFlux is working on some changes with respect to bundle loading

vargaz commented:
I think this is fine, the previous code was basically recursively collecting all the assemblies referenced by the app assembly, dotnet publish might do something slightly different.

radical commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+{
+    private List<string> _searchPaths;
+
+    public Resolver(List<string> searchPaths)
```
`IEnumerable<string>` should be enough, instead of using a concrete `List<string>` type.

mdh1418 replied:
Changed to use `string[]` to reflect `AssemblySearchPaths` type because of `_resolver = new Resolver(AssemblySearchPaths);`

radical commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+
+internal class Resolver : MetadataAssemblyResolver
+{
+    private List<string> _searchPaths;
```
`readonly`

radical commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+            }
+            catch (FileNotFoundException)
+            {
+            }
```
IIUC, this should also honor `SkipMissingAssemblies`. Also, log a message if skipping it.

radical commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+        {
+            try
+            {
+                Assembly refAssembly = mlc.LoadFromAssemblyName(aname);
```
This can also throw `BadImageFormatException`, and `FileLoadException`. This can do the same thing as the exception handler for `LoadFromAssemblyPath`.

mdh1418 replied:
How does `BadImageFormatException` get thrown?

radical replied:
Invalid assembly. Try with a simple text file.

radical commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+        return items;
+    }
+
+    private void Add(MetadataLoadContext mlc, Assembly assembly)
```
nit: `AddReferencedAssemblies` might be a more descriptive name.

mdh1418 replied:
Updated to be `AddAssemblyAndReferences` since it adds the assembly itself as well.

radical commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
```
Remove unused usings

radical commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public class WasmAssemblySearcher : Task
```
IMHO, the name doesn't indicate what this task really does. `WasmGetReferencedAssemblies`? or something better

mdh1418 replied:
How does `WasmLoadAssembliesAndReferences` sound?

radical commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+    }
+}
+
+internal class Resolver : MetadataAssemblyResolver
```
Maybe rename to be more descriptive, like `SearchPathsAssemblyResolver` (styled after `PathAssemblyResolver`)?

mdh1418 replied:
It applies to when AssemblySearchPaths is not available either, would `WasmAssemblyResolver` fare better?

radical replied:
Even in the other case, IIUC, it is still searching in a given "search path" (corelib's dir). It doesn't have anything specific to Wasm really.

radical commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+                var dir = v.ItemSpec;
+                if (!Directory.Exists(dir))
+                    throw new ArgumentException($"Directory '{dir}' doesn't exist or not a directory.");
+                paths.Add(dir);
```
Once you change `AssemblySearchPaths` to `string[]`, you can drop `paths`.

mdh1418 replied:
Thanks

radical commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+            {
+                foreach (var item in ExtraAssemblies)
+                {
+                    try
+                    {
+                        var refAssembly = mlc.LoadFromAssemblyPath(item.ItemSpec);
+                        Add(mlc, refAssembly);
+                    }
+                    catch (System.IO.FileLoadException)
+                    {
+                        if (!SkipMissingAssemblies)
+                            throw;
+                    }
```
Log a message if we are skipping the assembly. And we should catch other exceptions too:

```csharp
                    try
                    {
                        var refAssembly = mlc.LoadFromAssemblyPath(asmFullPath);
                        Add(mlc, refAssembly);
                    }
                    catch (Exception ex) when (ex is FileLoadException || ex is BadImageFormatException || ex is FileNotFoundException
                    {
                        if (SkipMissingAssemblies)
                        {
                            Log.LogMessage(MessageImportance.Low, $"Loading extra assembly '{asm}' failed with {ex}. Skipping");
                        }
                        else
                        {
                            Log.LogError($"Failed to load assembly from ExtraAssemblies '{asm}': {ex}");
                            return false;
                        }
                     }

```

mdh1418 replied:
It looks like `MetadataLoadContext.LoadFromAssemblyPath` only can throw `FileLoadException`, and any instance of `FileNotFoundException` arising from `Add` would be caught in the `Add` function.

Commit:
Clean up WasmAppBuilder if statements

Commit:
Readd WasmAppbuilder corelibPath logic

Commit:
Require non-null Assemblies in WasmAppBuilder

Commit:
Cleanup runtimeSourceDir logic

Commit:
Remove ExtraAssemblies from WasmAppBuilder

Commit:
Cleanup _assemblies in WasmAppBuilder

Commit:
Change Assemblies type in WasmAppBuilder

Commit:
Cleanup parameters

Commit:
Change parameter types

Commit:
Locally scope _resolver

Commit:
Address feedback

mdh1418 commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+    public ITaskItem[]? AssemblySearchPaths { get; set; }
+    public ITaskItem[]? Assemblies { get; set; }
+    public ITaskItem[]? ExtraAssemblies { get; set; }
+    public int DebugLevel { get; set; }
```
Thanks

mdh1418 commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+    [Output]
+    public ITaskItem[]? ReferencedAssemblies { get; set; }
+
+    private SortedDictionary<string, Assembly>? _assemblies;
```
Good catch

mdh1418 commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+    public ITaskItem[]? ReferencedAssemblies { get; set; }
+
+    private SortedDictionary<string, Assembly>? _assemblies;
+    private Resolver? _resolver;
```
Done

mdh1418 commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+        else
+        {
+            string corelibPath = string.Empty;
+            string runtimeSourceDir = string.Empty;
```
Thanks

Commit:
Make function name more descriptive

Commit:
Make Resolver function name more descriptive

Commit:
Rename task

radical commented on tools-local/tasks/mobile.tasks/WasmAppBuilder/WasmAssemblySearcher.cs:
```diff
@@ -0,0 +1,159 @@
+
+public class WasmLoadAssembliesAndReferences : Task
+{
+    public string? MainAssembly { get; set; }
+
+    // If true, continue when a referenced assembly cannot be found.
+    // If false, throw an exception.
+    public bool SkipMissingAssemblies { get; set; }
+
+    // Either one of these two need to be set
+    public string[]? AssemblySearchPaths { get; set; }
+    public string[]? Assemblies { get; set; }
+    public string[]? ExtraAssemblies { get; set; }
```
Actually, `MainAssembly` has no special significance here. This task is essentially taking bunch of assemblies as input, finding all the references to those, and returning the full set. And for resolving the references it has two cases:
1. specified AssemblySearchPaths
2. or, find `*Private.CoreLib.dll`, and look in it's containing directory

So, essentially, you can make this:
```CSharp
     // If true, continue when a referenced assembly cannot be found.
     // If false, throw an exception.
     public bool SkipMissingAssemblies { get; set; }

    // Either one of these two need to be set
    public string[]? AssemblySearchPaths { get; set; }

    [Required]
    [NotNull]
    public string[]? Assemblies { get; set; }
```

And in the Execute method, instantiate the `resolver`, depending on whether you have `AssemblySearchPaths` or not. And rest of the code to iterate over the assemblies should be the same. Something like:

```csharp
        foreach (var asm in Assemblies)
        {
            var asmFullPath = Path.GetFullPath(asm);
            try
            {
                var refAssembly = mlc.LoadFromAssemblyPath(asmFullPath);
                Add(mlc, refAssembly);
            }
            catch (Exception ex) when (ex is FileLoadException || ex is BadImageFormatException || ex is FileNotFoundException)
            {
                if (SkipMissingAssemblies)
                    Log.LogMessage(MessageImportance.Low, $"Loading extra assembly '{asm}' failed with {ex}. Skipping");
                else
                {
                    Log.LogError($"Failed to load assembly from ExtraAssemblies '{asm}': {ex}");
                    return false;
                }
            }
        }

```

radical replied:
And a follow up to that: msbuild already has a task that does this, it's called `ResolveAssemblyReference`, which can used here as:
```xml
    <ResolveAssemblyReference
      AssemblyFiles="bin\WasmSample.dll;$(MicrosoftNetCoreAppRuntimePackDir)lib\$(NetCoreAppCurrent)\System.Private.Runtime.InteropServices.JavaScript.dll"
      SearchPaths="@(AssemblySearchPaths)">
      <Output TaskParameter="ResolvedFiles" ItemName="ReferencedAssemblies" />
      <Output TaskParameter="ResolvedDependencyFiles" ItemName="ReferencedAssemblies" />
    </ResolveAssemblyReference>
```

.. instead of a new task. You could try this out. Check that you get the same list of assemblies as from *Searcher task.  If you find anything missing, then build with `/v:diag`, and that will show you why something didn't get resolved. https://docs.microsoft.com/en-us/visualstudio/msbuild/resolveassemblyreference-task?view=vs-2019

mdh1418 replied:
It does seem like there is redundant code that serve to iterate over a set of assemblies.
From what I understand, the two paths for the original task are:
1. If provided `AssemblySearchPaths`, ensure that all paths are valid, and then add the main assembly and its referenced assemblies. (From what I hear, `ExtraAssemblies` is a way to overcome the scenario where the linker removed essential assemblies which will be fixed when the real linker is used with `<PublishTrimmed>true</Publishtrimmed>`)
2. Without `AssemblySearchPaths`, its expected that `Assemblies` contains all essential assemblies to build the wasm app (of which, one assembly is expected to be `System.Private.CoreLib.dll`.

By refactoring the two cases into a standard process of 
```Figure out the right `_resolver` -> Iterate over a set of assemblies```
Then, if the `MainAssembly` and `ExtraAssemblies` parameters are removed now should need to be folded into `Assemblies` when `AssemblySearchPaths` != `null` outside the scope of the task.
Then is the expectation of the task to be of two forms
```
<WasmLoadAssembliesAndReferences
   ...
   AssemblySearchPaths="@(AssemblySearchPaths)"
   Assemblies="$(MainAssembly);@(ExtraAssemblies)" />
```
```
<WasmLoadAssembliesAndReferences
   ...
   <!-- No AssemblySearchPaths -->
   Assemblies="@(Assemblies)" />
```

radical replied:
Yeah, that's what I had in mind.

mdh1418 replied:
Looks like `ResolveAssemblyReference` is not loading some referenced assemblies. 
For `debugger-test.dll` which depends on `System.Console`, the resolved primary and reference assemblies only contains `debugger-test.dll` and `debugger-test.pdb`
For `WasmTestRunner.dll` which depends on atleast `System.Runtime`, the resolved reference assemblies are
```
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/Microsoft.DotNet.XHarness.TestRunners.Xunit.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Collections.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Private.CoreLib.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/Microsoft.DotNet.XHarness.TestRunners.Common.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/xunit.abstractions.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/xunit.runner.utility.netcoreapp10.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/xunit.execution.dotnet.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/xunit.core.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Private.Xml.Linq.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Private.Xml.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Linq.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Console.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.IO.FileSystem.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/Mono.Options.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Net.Sockets.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Collections.Concurrent.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Linq.Expressions.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.ObjectModel.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Diagnostics.TraceSource.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Text.RegularExpressions.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Net.Primitives.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Net.Http.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Collections.NonGeneric.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Collections.Specialized.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Runtime.CompilerServices.Unsafe.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Private.Uri.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Private.Runtime.InteropServices.JavaScript.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Runtime.InteropServices.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.ComponentModel.TypeConverter.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Security.Cryptography.Algorithms.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/Microsoft.Win32.Primitives.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Diagnostics.DiagnosticSource.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.ComponentModel.Primitives.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Drawing.Primitives.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.ComponentModel.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Runtime.Serialization.Formatters.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Security.Cryptography.Primitives.dll
/Users/mdhwang/runtime_wasm_clean/artifacts/bin/System.Runtime.Loader.Tests/net6.0-Release/browser-wasm/publish/System.Runtime.Numerics.dll
```
which doesn't contain `System.Runtime.dll`

radical replied:
The problem here is mismatched versions. The sample is referencing 5.0 assemblies, but we are trying to resolve them against 6.0 ones. `AutoUnify=true` helps with that, partially. I was working on a fix locally. I will try to clean that up, and share.

radical replied:
btw, you can run with `/v:diag`, and check the messages from `ResolveAssemblyReference` task, to see details of why it didn't resolve `System.Console`, which should show the version issue.

Commit:
Prevent copy of file to itself

Commit:
Clean up inconsistent formatting

radical commented:
Btw, make sure to run `make -C src/mono/wasm run-debugger-tests` too, because we are not running that on CI right now.

Commit:
Add asmRootPath to WasmAppBuilder

Commit:
Update debugger tests proj

Commit:
Default runtimeSource to runtime pack

Commit:
Add linker import and include WasmTestRunner dependencies

Commit:
Refactor logic to resolve assembly references

Commit:
Use ResolveAssemblyReference in WasmTestRunner

radical committed:
Fix the build, some cleanup, and additional error handling

radical committed:
WasmAppBuilder: Add exists checks for `File.Copy`, logging, and

.. correctly add to `@(FileWrites)`. Adding to that item is needed for
msbuild to be able to clean up.

radical committed:
add a missing check

radical commented:
I was running into some issues here, so I made some changes: https://github.com/mdh1418/runtime/compare/mdhwang/remove_wasm_app_mlc...radical:wasm-app

radical committed:
WasmLoadAssembliesAndReferences: Fail on missing input assemblies

`SkipMissingAssemblies` applies only to assembly references.

radical committed:
fix typo

radical committed:
WasmLoadAssembliesAndReferences: Remove magical mode

The task allows `AssemblySearchPaths` to not be set, in which case it
tries to find `System.Private.CoreLib.dll` in `Assemblies`, and uses
that path as a search path.

Remove that magical mode, and instead make `AssemblySearchPaths` a
required parameter. This means that the caller will find that correct
path. This makes the task clear in what it does.

radical committed:
WasmTestRunner: add missing UsingTask

Commit:
Publish untrimmed for suites with linker failures

Commit:
Redirect WasmFilesToIncludeInFileSystem and include System.Reflection.MetadataLoadContext missing assemblies

Commit:
Fix condition typo

Commit:
Untrim libraries tests

Commit:
SkipMissingAssemblies in WasmTestRunner

Commit:
Move WasmLoadAssembliesAndReferences to src/tasks

radical commented:
Would it be possible to split this into two parts?

1. Extract the new task from `WasmAppBuilder`, update all the projects files etc, but no behavior really changes. No new targets are imported etc.

2. Make the other changes needed to remove MLC bits, import linker targets etc.

The current state is kinda confusing to see what really changed other than the extraction of the task, and why.

steveisok commented on eng/testing/tests.mobile.targets:
```diff
@@ -157,28 +157,40 @@
-      <ExtraAssemblies Include="$(PublishDir)System.Drawing.dll" />
+      <Assemblies Include="$(PublishDir)\..\mscorlib.dll" />
+      <Assemblies Include="$(PublishDir)\..\System.Drawing.dll" />
+      <!-- WasmTestRunner referenced assemblies -->
```
You shouldn't need to explicitly call these out.  Would recommend removing.

mdh1418 replied:
There were errors regarding missing assemblies that `WasmTestRunner.dll` depended on because of the `<PublishTrimmed>true</PublishTrimmed>` addition. So I'll revert those changes long with removing `<PublishTrimmed>true</PublishTrimmed>`.

steveisok replied:
Yeah, we won't need that for the tests.  Too much pain.

mdh1418 commented:
For now, I think I'll remove the `<PublishTrimmed>` addition, which resulted in the extra assemblies being manually added.
Then I think all this PR will do is:
1.  extract the assembly and reference loading from the main `WasmAppBuilder`, thereby removing `WasmAppBuilder's` dependency on `System.Reflection.MetadataLoadContext` 
2. Modify the instances of `WasmAppBuilder` to account for the change in functionality.

Commit:
Revert manual assembly additions

steveisok commented on src/tasks/WasmAppBuilder/WasmAppBuilder.cs:
```diff
@@ -185,17 +151,21 @@ public override bool Execute ()
+        }
+        FileCopyChecked(MainJS!, Path.Join(AppDir, "runtime.js"), string.Empty);
 
         var html = @"<html><body><script type=""text/javascript"" src=""runtime.js""></script></body></html>";
```
Should creating index.html be optional?  Would it make sense to be able to supply one?

radical replied:
I'm not sure about it being optional. In what cases is it used, and not used?

But it would be helpful to be able to supply one. I have an upcoming PR which would be adding conditionally adding bits to the html.

mdh1418 replied:
I think for both samples in `dotnet/runtime` and `dotnet/samples`, the html gets overwritten/supplied in the csprojs.

steveisok commented on src/tasks/WasmAppBuilder/WasmAppBuilder.cs:
```diff
@@ -251,65 +221,43 @@ public override bool Execute ()
+        _fileWrites.Add(monoConfigPath);
 
-        using (var sw = File.CreateText(Path.Join(AppDir, "run-v8.sh")))
+        string runv8Path = Path.Join(AppDir, "run-v8.sh");
```
Does it make sense to write this file outside of the WasmAppBuilder and just pass it along as a property?  Something like `ExtraFiles` or whatever other name makes sense.

mdh1418 replied:
I'm not sure of the use case of `run-v8.sh`. Would it be used in a context that doesn't use WasmAppBuilder?

steveisok replied:
that's just a helper script to start, in our case, tests against v8.  I think it's an outside of WasmAppBuilder concern.

mdh1418 replied:
I'll put up this change in a follow up PR

radical commented on eng/testing/tests.mobile.targets:
```diff
@@ -157,16 +157,16 @@
+      <Assemblies Include="$(PublishDir)System.Drawing.dll" />
     </ItemGroup>
     <Error Condition="!Exists('$(MicrosoftNetCoreAppRuntimePackRidDir)')" Text="MicrosoftNetCoreAppRuntimePackRidDir=$(MicrosoftNetCoreAppRuntimePackRidDir) doesn't exist" />
     <WasmAppBuilder
```
Why is the new task not being used with this? IIUC, that would be needed to retain the existing "behavior" at least, and then a follow up PR can make the changes for streamlining that, or removing MLC etc.

radical replied:
And with that, you shouldn't need the change on line 160, or adding the `Import`

steveisok replied:
The reason why the new task exists is to make sure the runtime tests can still simulate a publish.  It's not needed anywhere else.  The libraries tests already publish.

radical replied:
I'm trying to understand why this is needed in here:
```diff
-      <AssemblySearchPaths Include="$(PublishDir)"/>
+       <Assemblies Include="$(PublishDir)\*.dll"/>
```

.. and
```diff
+ <Import Project="$(RepositoryEngineeringDir)illink.targets" Condition="'$(TargetOS)' == 'Browser'" />
```

steveisok replied:
We need to just carry over whatever is in the publish dir assemblies wise, so `<Assemblies Include="$(PublishDir)\*.dll"/>` is probably the only thing we need.  My guess about the others is that previously we were still going through MDLC and needed to force mscorlib, etc along.  

No idea about the linker import.  @mdh1418 ?

mdh1418 replied:
The import overrides the current linker being used so that it will use the linker from a nuget package rather than the one shipped from the dotnet sdk. The reason for this is that there was an error being thrown along the lines of `System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')`. The was already a fix for it but it's more recent and didn't yet make it to the current sdk being used. It was induced by adding `<PublishTrimmed>true</PublishTrimmed>` revealing the error, but we should still use it (regardless/independent from `<PublishTrimmed>`).

The original code will have resolved `MainAssembly` using paths provided in `AssemblySearchPaths` to build the WasmApp with necessary assemblies. However, these assemblies are now already resolved and placed in the `$(PublishDir)` as a result of publishing (I imagine it wasn't the case when `WasmAppBuilder` was first written). So now we can just include the assemblies necessary to run the app solely within the `Assemblies` parameter rather than needing to resolve them again (which needs to be done for the debugger-tests and runtime tests since they do not publish).

radical replied:
>The original code will have resolved MainAssembly using paths provided in AssemblySearchPaths to build the WasmApp with necessary assemblies. However, these assemblies are now already resolved and placed in the $(PublishDir) as a result of publishing (I imagine it wasn't the case when WasmAppBuilder was first written). So now we can just include the assemblies necessary to run the app solely within the Assemblies parameter rather than needing to resolve them again (which needs to be done for the debugger-tests and runtime tests since they do not publish).

Great! Can you include this in final commit message/PR description too?

> The was already a fix for it but it's more recent and didn't yet make it to the current sdk being used. It was induced by adding <PublishTrimmed>true</PublishTrimmed> revealing the error, but we should still use it (regardless/independent from <PublishTrimmed>).

I'm not sure I understand why it needs to be added here. IIUC, it is not being enabled here, and `illink.targets` should not be needed here.

mdh1418 replied:
Done, modified a bit so it would be a bit less wordy :)

Commit:
Remove nuget linker import

In the libraries test, tests.mobile.targets, WasmAppBuilder originally resolved and loaded the MainAssembly using paths provided in AssemblySearchPaths to build the WasmApp. However, the libraries test now uses publish, which does that already. So we now include the assemblies resulting from publish directly into WasmAppBuilder in order to run the libraries test. The debugger-tests and runtime tests do not use publish yet, so they will use the new task WasmLoadAssembliesAndReferences to resolve necessary assemblies.

radical reviewed (approved):
Thank you for your patience, and quick responses to feedback!
LGTM 👍
